src/tasks.coffee

```diff
@@ -0,0 +1,314 @@
+    else if task
+      activeTasks++
+      processNextTaskRound task, (err) ->
+        logger.err "An error occurred while processing rerun task #{task._id}: #{err}" if err
+        activeTasks--
+        # task has finished its current round, pick up the next one
+        if live then findAndProcessAQueuedTask()
+
+      # task has started processing, pick up the next one
+      if live then findAndProcessAQueuedTask()
+
+rerunTaskProcessor = ->
+  if live
+    findAndProcessAQueuedTask()
+    setTimeout rerunTaskProcessor, config.rerun.processor.pollPeriodMillis
+
+
+exports.start = (callback) ->
+  live = true
+  setTimeout rerunTaskProcessor, config.rerun.processor.pollPeriodMillis
+
+  logger.info "Started rerun task processor"
+  callback()
+
+
+exports.stop = (callback) ->
+  live = false
+
+  waitForActiveTasks = ->
+    if activeTasks > 0
+      setTimeout waitForActiveTasks, 100
+    else
+      logger.info "Stopped rerun task processor"
+      callback()
+
+  waitForActiveTasks()
+
+exports.isRunning = -> live
+
+
+finalizeTaskRound = (task, callback) ->
+  # get latest status in case it has been changed
+  TaskModel.findOne { _id: task._id }, { status: 1 }, (err, result) ->
+    return callback err if err
+
+    # Only queue the task if still in 'Processing'
+    # (the status could have been changed to paused or cancelled)
+    if result.status is 'Processing' and task.remainingTransactions isnt 0
+      task.status = 'Queued'
+      logger.info "Round completed for rerun task ##{task._id} - #{task.remainingTransactions} transactions remaining"
+    else
+      if task.remainingTransactions is 0
+        task.status =  'Completed'
+        task.completedDate = new Date()
+        logger.info "Round completed for rerun task ##{task._id} - Task completed"
+      else
+        task.status = result.status
+        logger.info "Round completed for rerun task ##{task._id} - Task has been #{result.status}"
+
+    task.save (err) -> callback err
+
+# Process a task.
+#
+# Tasks are processed in rounds:
+# Each round consists of processing n transactions where n is between 1 and the task's batchSize,
+# depending on how many transactions are left to process.
+#
+# When a round completes, the task will be marked as 'Queued' if it still has transactions remaining.
+# The next available core instance will then pick up the task again for the next round.
+#
+# This model allows the instance the get updated information regarding the task in between rounds:
+# i.e. if the server has been stopped, if the task has been paused, etc.
+processNextTaskRound = (task, callback) ->
+  promises = []
+  nextI = task.transactions.length - task.remainingTransactions
+
+  for transaction in task.transactions[nextI ... nextI+task.batchSize]
+    do (transaction) ->
+      defer = Q.defer()
+
+      rerunTransaction transaction.tid, task._id, (err, response) ->
+        if err
+          transaction.tstatus = 'Failed'
```

Comment:
No, we're storing more than just the tid (https://github.com/jembi/openhim-core-js/blob/master/src/model/tasks.coffee#L8), we're storing things like the transaction process status and result. The transaction object is part of the task object, so these updates take effect [here](https://github.com/jembi/openhim-core-js/blob/tasks-batching-strategy/src/tasks.coffee#L80) and [here](https://github.com/jembi/openhim-core-js/blob/tasks-batching-strategy/src/tasks.coffee#L120)